src/Eventso.Subscription.Kafka.DeadLetter.Postgres/PoisonEventRetryScheduler.cs

```diff
@@ -0,0 +1,104 @@
+using Confluent.Kafka;
+using Npgsql;
+
+namespace Eventso.Subscription.Kafka.DeadLetter.Postgres;
+
+internal sealed class PoisonEventRetryScheduler(
+    IConnectionFactory connectionFactory,
+    int maxAllowedFailureCount,
+    TimeSpan minIntervalBetweenRetries,
+    TimeSpan maxLockHandleInterval)
+    : IPoisonEventRetryScheduler
+{
+    public async Task<ConsumeResult<byte[], byte[]>?> GetNextRetryTarget(string groupId, TopicPartition topicPartition, CancellationToken token)
+    {
+        await using var connection = connectionFactory.ReadOnly();
+
+        await connection.OpenAsync(token);
+
+        await using var command = new NpgsqlCommand(
+            // language=sql
+            """
+            WITH
+                heads AS (
+                    SELECT key, MIN("offset") AS min_offset
+                    FROM eventso_dlq.poison_events
+                    WHERE topic = @topic AND partition = @partition AND group_id = @groupId
+                    GROUP BY key
+                ),
+                locked_events AS (
+                    SELECT pe_heads.*
+                    FROM eventso_dlq.poison_events pe_heads
+                    INNER JOIN heads
+                    ON pe_heads.topic = @topic
+                        AND pe_heads.partition = @partition
+                        AND pe_heads."offset" = heads.min_offset
+                        AND pe_heads.key = heads.key
+                        AND pe_heads.group_id = @groupId
+                    WHERE
+                        pe_heads.total_failure_count <= @maxFailureCount
+                        AND pe_heads.last_failure_timestamp < @maxLastFailureTimestamp
+                        AND (pe_heads.lock_timestamp IS NULL OR pe_heads.lock_timestamp < @maxLockTimestamp)
+                    FOR UPDATE OF pe_heads SKIP LOCKED
+                    LIMIT 1
```

Comment:
I'd suggest smarter explicit sorting of candidate heads, i.e. unlocked, lowest retries, oldest keys first


```suggestion
                    ORDER BY pe_heads.lock_timestamp ASC NULLS FIRST, pe_heads.total_failure_count ASC, pe_heads.last_failure_timestamp ASC
                    FOR UPDATE OF pe_heads SKIP LOCKED
                    LIMIT 1
```

Reply:
interesting idea, ++